src/main/java/org/zalando/nakadi/validation/schema/diff/SchemaDiff.java

```diff
@@ -29,25 +30,39 @@
     }
 
     static void recursiveCheck(
-            final Schema original,
-            final Schema update,
+            final Schema originalIn,
+            final Schema updateIn,
             final SchemaDiffState state) {
 
-        if (original == null && update == null) {
+        if (originalIn == null && updateIn == null) {
             return;
         }
 
-        if (update == null) {
+        if (updateIn == null) {
             state.addChange(SCHEMA_REMOVED);
             return;
         }
 
-        if (!original.getClass().equals(update.getClass())) {
-            state.addChange(TYPE_CHANGED);
-            return;
+        final Schema original;
+        final Schema update;
+        if (!originalIn.getClass().equals(updateIn.getClass())) {
+            // Tricky part. EmptySchema is the same as an empty ObjectSchema.
+            if (originalIn instanceof EmptySchema && updateIn instanceof ObjectSchema) {
+                original = replaceWithEmptyObjectSchema(originalIn);
+                update = updateIn;
+            } else if (originalIn instanceof ObjectSchema && updateIn instanceof EmptySchema) {
```

Comment:
Is this correct? If I understood it correctly, this would allow changing:

```javascript
{
  "properties": {
    "foo": {
      "type":"string"
    }
  }
}
```

to

```javascript
{
}
```

The first schema does not match the event `[]` but the later does.

I think that lines 48 to 52 are correct, but I'm not sure about 53 to 55.

Reply:
@rcillo Yes, this is correct. The purpose of this 2 steps are to replace EmptySchema with empty ObjectSchema, later it will call usual process of validation. 
In the case you pointed it will actually compare 2 schemas: 
```
{
  "properties": {
    "foo": {
      "type":"string"
    }
  }
}
```
and 
```
{
  "properties": {}
}
```
And the possibility of this change is regulated by ObjectSchema checks.

Reply:
@antban ok, now I got it. So in this case it would reject the change because the user would have removed one key.

But then I think there is another problem. It would accept the change:

```javascript
{
  "properties": {}
}
```

To

```javascript
{}
```

This change should not be allowed because the second schema accepts and array event `[]` as valid but the former doesn't. 

Makes sense?

Reply:
@rcillo Yep, makes sense. Changed this behavior: https://github.com/zalando/nakadi/pull/829/commits/91431cbe835f3bb433b05ff21ea4043745784f7f